launch_handler-explainer.md

```diff
@@ -0,0 +1,141 @@
+# Web App Launch Handling
+
+## Overview
+
+This document describes a new `launch_handler` manifest member that enables
+web apps to customise their launch behaviour across all types of app launch
+triggers.
+
+
+## Use Cases
+
+- Web apps that are designed to be used in a single window e.g. a music app.
+
+- Web app that capture and handle share target events and user navigations
+  in existing windows without invoking a navigation and losing existing state.
+  E.g. sharing an image to a chat web app could open a contact picker overlayed
+  on top of existing chat content.
+
+
+## Background
+
+There are several ways for a web app window to be opened:
+- [File handling](https://github.com/WICG/file-handling/blob/main/explainer.md)
+- [In scope link capturing](https://github.com/WICG/sw-launch/blob/master/declarative_link_capturing.md)
+- [Note taking](https://wicg.github.io/manifest-incubations/index.html#note_taking-member)
+- Platform specific app launch surface
+- [Protocol handling](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/URLProtocolHandler/explainer.md)
+- [Share target](https://w3c.github.io/web-share-target/)
+- [Shortcuts](https://www.w3.org/TR/appmanifest/#dfn-shortcuts)
+
+Web apps launched via these triggers will open in a new or existing app window
+depending on the user agent platform. There is currently no mechanism for the
+web app to configure this behaviour.
+
+
+## Proposal
+
+- Add a `launch_handler` member to the web app manifest specifying the default
+  client selection and navigation behaviour for web app launches.
+  The shape of this member is as follows:
+  ```
+  "launch_handler": {
+    "route_to": "auto" | "new" | "existing",
+    "navigate_client": "auto" | true | false
+  }
+  ```
+
+  If unspecified then `launch_handler` defaults to
```

Comment:
I don't think leaving it up to the user agent is good. Usually, this is done for three reasons:

- Because it's browser UI that doesn't concern the site developer, or
- Because it should be the user's choice, or
- For historical reasons (browsers diverge and there's no consensus).

None applies here. This does concern the site developer because it changes what events they need to be ready for. For example, `"navigate_client": "auto"` technically means the browser can open the unsuspecting app without navigating it, and fire a new event that the site hasn't heard of. This totally breaks a site that hasn't explicitly opted into it. `"route_to": "auto"` is a little less intrusive -- you won't completely break the site if you choose to navigate an existing window, but I still think this is something we shouldn't do without the site's opt-in (which is _why_ when we launched Desktop PWAs in Chrome in 2018, we decided _not_ to apply this behaviour).

The default behaviour should be prescribed as `"new"` + `true`, which most closely matches normal HTML navigation. And that means we don't need an explicit `"auto"` value.

**However** that's not really right either. It seems that the intended implementation here is that `{"route_to": "auto"}` disables link capturing, whereas `{"route_to": "new"}` enables it an opens a new window. (Is that right?) In which case, `"auto"` is a third, entirely different value; so it should be called `"none"`, not `"auto"` (and it should be the default).

**HOWEVER**, really this last point applies only to *link capturing*, not other forms of launching like the home screen icon, shortcuts, share targets, etc, which would always open a window. So in fact, this is a third orthogonal choice. So there actually should be a third value here, called ... tada ... `"capture_links"`, which is a simple Boolean (or equivalent enum) that says whether clicking a link should activate this whole launch handler mechanism. Other forms of launching would ignore `capture_links`.

You could also put `capture_links` in the top level, outside of `launch_handler`, which makes it a peer of other things that activate the launch handler, like `share_target`, `shortcuts`, etc. I think that is better.

So we're left with:

```js
"launch_handler": {
  "route_to": "new" | "existing",  // default = "new"
  "navigate_client": true | false  // default = "true"
},
"capture_links": true | false  // default = false
```

The default value aligns exactly with what we have today, in a pre-launch-handler world.

Reply:
Good point re `"navigate_client": "auto"`; it doesn't make a whole lot of sense and should default to true/always instead.

`"route_to": "auto"` is intended to capture the state of the world today where on mobile it can only be "existing-client" while on desktop it should default to "new-client" to avoid state loss.

Re `capture_links`: link capturing is an orthogonal behaviour that decides whether or not a link navigation should trigger an app launch (like how share target and file handlers trigger app launches via other user actions). I intend for `launch_handler` to come into effect only after a launch has been triggered (link capturing or otherwise) and have no influence on the launch triggers themselves.

Re link capturing: I don't think we should block users from enabling link capturing for web apps nor should we enable link capturing without the user's opt in. This makes link capturing behaviour a user decision configured in the user agent rather than the site manifest.